### PR TITLE
Inital work on refactoring neaby into a service.

### PR DESCRIPTION
**Pull request by xiamaz**
@zerik Could you help me on getting a notification to show inside the nearby service? I will concentrate on getting the connection to work today for events.

### CODE REVIEW

**Comment by ghost** [comment]
@xiamaz it seems to me that showing the notification is already implemented. But I can't test whether it works because I don't have a test device.

**Comment by xiamaz** [comment]
It didnt work for me in VM, so it would be great if we could debug it further.

On 12 December 2018 21:42:47 CET, zerik <notifications@github.com> wrote:
>@xiamaz it seems to me that showing the notification is already
>implemented. But I can't test whether it works because I don't have a
>test device.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/amos-flyinn/amos-tub-ws18-proj1/pull/48#issuecomment-446736557

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by xiamaz** [comment]
Tests have been written for nearbyservice on the app side. The server side still uses nearby directly inside the main activity. This seems reasonable in this case, since an activity will always be running.

File: app/src/main/java/com/amos/flyinn/ShowCodeActivity.java [hunk 1]
```diff
@@ -1,40 +1,117 @@
...
-import java.util.List;
 import java.util.concurrent.ThreadLocalRandom;
 
-
-public class ShowCodeActivity extends WifiConnectorBase {
+/**
+ * Initial activity showing code used for connection from remote display.
+ */
+public class ShowCodeActivity extends AppCompatActivity {
     private String nameNum = "1234";
```
Comment (lbb) [inline app/src/main/java/com/amos/flyinn/ShowCodeActivity.java]: Can we make this random?

File: app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    /**
+     * Create a notification channel.
+     *
+     * Notifications are organized into different channels to theoretically enable the user to
+     * individually set what they want to be informed about.
+     *
+     * This is required since Android 8.
+     */
+    private void createChannel() {
+        NotificationManager mgr=
```
Comment (lbb) [inline app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java:81]: Formatting?

File: app/src/test/java/com/amos/flyinn/nearbyservice/NearbyServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        service.setServiceState(NearbyState.CONNECTED, "Test");
+        assertEquals(NearbyState.CONNECTED, service.getServiceState());
+        service.setServiceState(NearbyState.CONNECTING, "Test");
+        assertEquals(NearbyState.CONNECTING, service.getServiceState());
+    }
+
+    @Test
+    public void testNotify() {
+        Intent intent = new Intent(ApplicationProvider.getApplicationContext(), TestService.class);
+        intent.putExtra("code", "1234");
```
Comment (lbb) [inline app/src/test/java/com/amos/flyinn/nearbyservice/NearbyServiceTest.java:112]: Do we want to leave this static? Can we test it with 9999 ?

File: app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+    /**
+     * Set state of the nearby service. This is used by the nearby server.
+     * @param state
+     * @param message
+     */
+    public void setServiceState(NearbyState state, @Nullable String message) {
+        // do extra things if we are switching state
+        if (serviceState != state) {
+            switch(serviceState) {
```
Comment (lbb) [inline app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java:177]: Why do we need a switch?

Comment (xiamaz) [reply app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java:177]: This has been implemented as a switch to handle finer reporting and handling of connection states later on. Disconnection and reconnect behavior is still quite weird

File: app/src/test/java/com/amos/flyinn/nearbyservice/NearbyServiceTest.java (hunk 3 above)
Comment (xiamaz) [reply app/src/test/java/com/amos/flyinn/nearbyservice/NearbyServiceTest.java:112]: There are no bounds checks in the code setter in the server activity. IMO it makes more sense if we are actually defining more restrictive behavior. At the current stage they all are basically checks, whether stuff will fail catastrophically

File: app/src/main/java/com/amos/flyinn/ShowCodeActivity.java (hunk 1 above)
Comment (xiamaz) [reply app/src/main/java/com/amos/flyinn/ShowCodeActivity.java]: It is set randomly in oncreate.

File: app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java (hunk 2 above)
Comment (xiamaz) [reply app/src/main/java/com/amos/flyinn/nearbyservice/NearbyService.java:81]: Fixed

File: app/src/main/java/com/amos/flyinn/ShowCodeActivity.java (hunk 1 above)
Comment (xiamaz) [reply app/src/main/java/com/amos/flyinn/ShowCodeActivity.java]: Also changed default to something more apparent.

**Comment by xiamaz** [comment]
@lbb PTAL

**Review by lbb — APPROVED** [review]
LGTM

**Pull request merged**